### DISCUSSION TITLE
FastAPI LifeSpan is delayed when using Docker Compose

### DISCUSSION THREAD

**Discussion by MuhammadSalah99** [fastapi/fastapi#12751] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@asynccontextmanager
async def lifespan(app: FastAPI):
    print("start")
    yield
    print("bye from start")

main.py:
app = FastAPI(lifespan = lifespan)


@app.get("/")
async def root():
    return {"message": "Hello World"}

docker-compose.yml
fastapi:
    build:
      context: ./backend   
      dockerfile: Dockerfile 
    volumes:
      - ./data:/app/data  
    working_dir: /app
    ports:
      - "8000:8000"
    environment:
      DATABASE_URL: "mysql+pymysql://user:admin123@mysql:3306/price_benchmarking"
    depends_on:
      mysql:
        condition: service_healthy
```

### Description

Hi 
I am using Docker Compose to run my application with MySQL v8.

when I run `docker compose up --build`
I get this
`
fastapi-1  | INFO:     Waiting for application startup.
fastapi-1  | INFO:     Application startup complete.
`
without printing anything, but as soon as I hit a request it prints out the value I have in my lifespan function.

please help

### Operating System

Linux

### FastAPI Version

0.115.4

### Pydantic Version

2.9.2

### Python Version

Python 3.10.12

**Comment by YuriiMotov** [comment]
Just checked it and I can't confirm such behavior.
In my case I can see `start` message before any other log messages from the container:

```
 ✔ Container proj-fastapi-1  Recreated                                                                                                                                                                  0.1s 
Attaching to fastapi-1
fastapi-1  | start
fastapi-1  | INFO:     Started server process [1]
fastapi-1  | INFO:     Waiting for application startup.
fastapi-1  | INFO:     Application startup complete.
fastapi-1  | INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
```

**Reply by MuhammadSalah99** [reply]
can you please share your Dockerfile with me


I have this CMD instruction in mine, ` CMD ["uvicorn", "main:app", "--host", "0.0.0.0", "--port", "8000", "--reload"]` 

I forgot to include it before.

could the issue be from it?

**Reply by YuriiMotov** [reply]
Let's better you share your Dockerfile and we will try to understand what might be wrong with it?
Also, please provide log history from the moment container started till you finally see the `start` message

**Reply by MuhammadSalah99** [reply]
thanks for replying

Dockerfile:
```

`
FROM python:3.9

WORKDIR /app

COPY requirements.txt .

RUN pip install  -r requirements.txt

COPY . .

EXPOSE 8000

CMD ["uvicorn", "main:app", "--host", "0.0.0.0", "--port", "8000", "--reload"]

`

```
here is the full log history:
```

 INFO:     Will watch for changes in these directories: ['/app']
fastapi-1   | INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
fastapi-1   | INFO:     Started reloader process [1] using StatReload
fastapi-1   | INFO:     Started server process [8]
fastapi-1   | INFO:     Waiting for application startup.
fastapi-1   | INFO:     Application startup complete.
```

when I hit an endpoint on the backend, this will be printed beneath the "Application Startup complete."

```
fastapi-1   | start
fastapi-1   | INFO:     172.18.0.1:53344 - "GET / HTTP/1.1" 200 OK
```


so it didn't print untill we hit the endpoint.

**Reply by YuriiMotov** [reply, marked as answer]
Yes, with this Dockerfile I also see `start` message only when I hit the endpoint.
But actually, your lifespan is executed on app startup. The problem is just with log output.
I guess that's because you are using `print`. This message is being stored in the ouput buffer, but is actually being printed only with next log output command.

So, if you just add the `logger.info` inctruction after your `print` (and configure the logger) it will be printed immediately:
```
from contextlib import asynccontextmanager
import logging
import sys

from fastapi import FastAPI

logger = logging.getLogger(__name__)
logging.basicConfig(
    level=logging.INFO,
    format="%(message)s",
    handlers=[logging.StreamHandler(sys.stdout)],
)

@asynccontextmanager
async def lifespan(app: FastAPI):
    print("starting")
    logger.info("start")
    yield
    logger.info("bye from start")


app = FastAPI(lifespan = lifespan)


@app.get("/")
async def root():
    return {"message": "Hello World"}

```

As a conclusion: just use logging instead of `print`s!
